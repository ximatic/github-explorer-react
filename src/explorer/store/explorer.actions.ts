export enum ExplorerAction {
  Reset = 'explorer/reset',
  //ResetToken = "explorer/resetToken",
  VerifyToken = 'explorer/verifyToken',
  VerifyTokenSuccess = 'explorer/verifyTokenSuccess',
  VerifyTokenError = 'explorer/verifyTokenError',
  //Token = 'explorer/token',
  LoadRepositories = 'explorer/loadRepositories',
  LoadRepositoriesSuccess = 'explorer/loadRepositoriesSuccess',
  LoadRepositoriesError = 'explorer/loadRepositoriesError',
  LoadRepository = 'explorer/loadRepository',
  LoadRepositorySuccess = 'explorer/loadRepositorySuccess',
  LoadRepositoryError = 'explorer/loadRepositoryError',
  PageInfo = 'explorer/pageInfo',
}
