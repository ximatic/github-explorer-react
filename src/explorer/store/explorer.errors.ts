export enum ExplorerError {
  InvalidToken = 'EXPLORER_ERROR_INVALID_TOKEN',
}
