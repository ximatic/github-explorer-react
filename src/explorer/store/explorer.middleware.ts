import { Middleware } from '@reduxjs/toolkit';

import { ApolloQueryResult, gql } from '@apollo/client';

import { apolloClient } from '../../ApolloProvider';

import { GraphqlPageInfo, GraphqlRepositoriesData, GraphqlRepositoriesNode } from '../models/explorer-graphql.model';
import { ExplorerPageInfo, Repository } from '../models/explorer.model';

import { ExplorerAction } from './explorer.actions';
import { loadRepositoriesError, loadRepositoriesSuccess, pageInfo, verifyTokenError, verifyTokenSuccess } from './explorer.slice';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const explorerMiddleware: Middleware = (storeAPI: any) => (next: any) => (action: any) => {
  const result = next(action);

  const storageKey = 'gh-explorer-token';

  // token

  function storeToken(token: string): void {
    localStorage.setItem(storageKey, JSON.stringify(token));
  }

  function resetToken(): void {
    localStorage.removeItem(storageKey);
  }

  // page info

  function createPageInfo(itemsPerPage: number, pageInfo: GraphqlPageInfo): ExplorerPageInfo {
    return {
      itemsPerPage,
      cursorStart: pageInfo.startCursor,
      cursorEnd: pageInfo.endCursor,
      hasNextPage: pageInfo.hasNextPage,
      hasPreviousPage: pageInfo.hasPreviousPage,
    };
  }

  switch (action.type) {
    case ExplorerAction.Reset:
      resetToken();
      break;
    case ExplorerAction.VerifyToken:
      try {
        const token = action.payload.token;
        fetch('https://api.github.com/octocat', {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        })
          .then(async (data: Response) => {
            if (!data.ok) {
              storeAPI.dispatch(verifyTokenError());
              return;
            }
            if (action.payload.storeToken) {
              storeToken(token);
            }
            storeAPI.dispatch(verifyTokenSuccess({ token }));
          })
          .catch(() => {
            storeAPI.dispatch(verifyTokenError());
          });
      } catch {
        storeAPI.dispatch(verifyTokenError());
      }
      break;
    case ExplorerAction.LoadRepositories:
      try {
        const pagination = action.payload.pagination;
        const token = storeAPI.getState().explorer.token;

        apolloClient
          .query({
            query: gql`
                query {
                  search(query: "is:public sort:stars-desc", type: REPOSITORY, ${pagination.itemsKey}: ${pagination.itemsValue}, ${pagination.cursorKey}: ${pagination.cursorValue}) {
                    nodes {
                      ...repositoryFragment
                    }
                    pageInfo {
                      endCursor
                      startCursor
                      hasNextPage
                      hasPreviousPage
                    }
                  }
                }
                fragment repositoryFragment on Repository {
                  name
                  createdAt
                  owner {
                    login
                  }
                  description
                  stargazerCount
                }
              `,
            context: {
              headers: {
                Authorization: `Bearer ${token}`,
              },
            },
          })
          .then((result: ApolloQueryResult<GraphqlRepositoriesData>) => {
            const repositories = result.data.search.nodes.map((node: GraphqlRepositoriesNode) => {
              return {
                name: node.name,
                owner: node.owner?.login || 'N/A',
                stars: node.stargazerCount,
                createdAt: node.createdAt,
                description: node.description,
              } as Repository;
            });

            storeAPI.dispatch(loadRepositoriesSuccess({ repositories }));
            storeAPI.dispatch(pageInfo({ pageInfo: createPageInfo(pagination.itemsValue, result.data.search.pageInfo) }));
          });
      } catch {
        storeAPI.dispatch(loadRepositoriesError());
      }
      break;
  }

  return result;
};
