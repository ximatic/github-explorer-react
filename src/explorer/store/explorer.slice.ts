import { createSlice, ReducerCreators, type PayloadAction } from '@reduxjs/toolkit';

import { ExplorerEvent, ExplorerEventName, ExplorerEventType, ExplorerState } from './explorer.state';
import { ExplorerPageInfo, ExplorerPagination, Repository } from '../models/explorer.model';

const initialState: ExplorerState = {
  token: null,
  repositories: null,
  repository: null,
  pageInfo: null,
  error: null,
  event: null,
};

export const explorerSlice = createSlice({
  name: 'explorer',
  initialState: initialState,
  reducers: (create: ReducerCreators<ExplorerState>) => ({
    // reset
    reset: create.reducer(() => initialState),
    // verify token
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    verifyToken: create.reducer((state: ExplorerState, action: PayloadAction<{ token: string; storeToken: boolean }>) => {
      state.token = initialState.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Processing,
      };
    }),
    verifyTokenSuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ token: string }>) => {
      state.token = action.payload.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Success,
      };
    }),
    verifyTokenError: create.reducer((state: ExplorerState) => {
      state.token = initialState.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Error,
      };
    }),
    // load repositories
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    loadRepositories: create.reducer((state: ExplorerState, action: PayloadAction<{ pagination: ExplorerPagination }>) => {
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Processing,
      };
    }),
    loadRepositoriesSuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ repositories: Repository[] }>) => {
      state.repositories = action.payload.repositories;
      state.repository = initialState.repository;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Success,
      };
    }),
    loadRepositoriesError: create.reducer((state: ExplorerState) => {
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Error,
      };
    }),
    // load repository
    loadRepository: create.reducer(
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      (state: ExplorerState, action: PayloadAction<{ owner: string; name: string; pagination: ExplorerPagination }>) => {
        state.pageInfo = initialState.pageInfo;
        state.event = {
          name: ExplorerEventName.LoadRepository,
          type: ExplorerEventType.Processing,
        };
      },
    ),
    loadRepositorySuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ repository: Repository }>) => {
      state.repository = action.payload.repository;
      state.event = {
        name: ExplorerEventName.LoadRepository,
        type: ExplorerEventType.Success,
      };
    }),
    loadRepositoryError: create.reducer((state: ExplorerState) => {
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepository,
        type: ExplorerEventType.Error,
      };
    }),
    // pagination
    pageInfo: create.reducer((state: ExplorerState, action: PayloadAction<{ pageInfo: ExplorerPageInfo }>) => {
      state.pageInfo = action.payload.pageInfo;
    }),
  }),
  selectors: {
    selectExplorerEvent: (state: ExplorerState): ExplorerEvent | null => state.event,
    selectToken: (state: ExplorerState): string | null => state.token,
    selectRepositories: (state: ExplorerState): Repository[] | null => state.repositories,
    selectRepository: (state: ExplorerState): Repository | null => state.repository,
    selectPageInfo: (state: ExplorerState): ExplorerPageInfo | null => state.pageInfo,
  },
});

export const {
  reset,
  verifyToken,
  verifyTokenSuccess,
  verifyTokenError,
  loadRepositories,
  loadRepositoriesSuccess,
  loadRepositoriesError,
  loadRepository,
  loadRepositorySuccess,
  loadRepositoryError,
  pageInfo,
} = explorerSlice.actions;

export const { selectExplorerEvent, selectToken, selectRepositories, selectRepository, selectPageInfo } = explorerSlice.selectors;
