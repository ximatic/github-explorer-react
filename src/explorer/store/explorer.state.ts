import { Repository, ExplorerPageInfo } from '../models/explorer.model';
import { ExplorerError } from './explorer.errors';

export enum ExplorerEventName {
  VerifyToken = 'verify-token',
  LoadRepositories = 'load-repositories',
  LoadRepository = 'load-repository',
}

export enum ExplorerEventType {
  Processing = 'processing',
  Success = 'success',
  Error = 'error',
}

export interface ExplorerEvent {
  name: ExplorerEventName;
  type: ExplorerEventType;
  message?: string;
}

export interface ExplorerState {
  token: string | null;
  repositories: Repository[] | null;
  repository: Repository | null;
  pageInfo: ExplorerPageInfo | null;
  error: ExplorerError | null;
  event: ExplorerEvent | null;
}
