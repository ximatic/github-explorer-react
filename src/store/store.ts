import { configureStore } from '@reduxjs/toolkit';

import { explorerSlice } from '../explorer/store/explorer.slice';
import { explorerMiddleware } from '../explorer/store/explorer.middleware';

// export const store = configureStore({
//   reducer: {
//     explorer: explorerSlice.reducer,
//   },
//   middleware: (getDefaultMiddleware) => {
//     return getDefaultMiddleware().concat(explorerMiddleware);
//   },
// });

// export type AppStore = typeof store;
// export type AppDispatch = typeof store.dispatch;
// export type RootState = ReturnType<typeof store.getState>;

export const makeStore = () => {
  return configureStore({
    reducer: {
      explorer: explorerSlice.reducer,
    },
    middleware: (getDefaultMiddleware) => {
      return getDefaultMiddleware().concat(explorerMiddleware);
    },
  });
};

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];
