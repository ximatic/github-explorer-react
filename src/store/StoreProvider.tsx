'use client';

import { useRef } from 'react';
import { Provider } from 'react-redux';

import { AppStore, makeStore } from './store';

interface StoreProviderProps {
  readonly children: React.ReactNode;
}

export default function StoreProvider({ children }: StoreProviderProps) {
  const storeRef = useRef<AppStore | null>(null);
  if (!storeRef.current) {
    // Create the store instance the first time this renders
    storeRef.current = makeStore();
  }

  return <Provider store={storeRef.current}>{children}</Provider>;
}
