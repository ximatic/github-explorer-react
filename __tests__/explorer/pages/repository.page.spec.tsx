import { render, screen, fireEvent } from '@testing-library/react';

import { BrowserRouter, MemoryRouter, Routes, Route } from 'react-router-dom';

import { PrimeReactProvider } from 'primereact/api';

import {
  MOCK_EVENT_LOAD_REPOSITORY_ERROR,
  MOCK_EXPLORER_PAGE_INFO_1,
  MOCK_REPOSITORY_1,
  MOCK_REPOSITORY_ISSUE_1,
  MOCK_REPOSITORY_ISSUE_2,
} from '../../../__mocks__/explorer.mocks';

import { useAppDispatch, useAppSelector } from '../../../src/store/hooks';
import {
  loadRepository,
  selectExplorerEvent,
  selectPageInfo,
  selectRepository,
} from '../../../src/explorer/store/explorer.slice';
import { defaultExplorerPagination } from '../../../src/explorer/models/explorer.model';

import RepositoryPage from '../../../src/explorer/pages/repository.page';

jest.mock('../../../src/store/hooks');

const mockUseAppDispatch = useAppDispatch as unknown as jest.Mock;
const mockUseAppSelector = useAppSelector as unknown as jest.Mock;

describe('RepositoryPage', () => {
  let dispatchMock: jest.Mock;

  const renderComponent = () =>
    render(
      <BrowserRouter>
        <PrimeReactProvider>
          <RepositoryPage />
        </PrimeReactProvider>
      </BrowserRouter>,
    );

  const renderComponentWithRoute = (route: string) =>
    render(
      <MemoryRouter initialEntries={[route]}>
        <PrimeReactProvider>
          <Routes>
            <Route path='/:owner/:name' element={<RepositoryPage />} />
          </Routes>
        </PrimeReactProvider>
      </MemoryRouter>,
    );

  beforeEach(() => {
    dispatchMock = jest.fn();

    mockUseAppDispatch.mockReturnValue(dispatchMock);
    mockUseAppSelector.mockReturnValue(undefined);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  // snapshot tests

  test('matches snapshot', () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  // unit tests

  it('deos not dispatch loadRepository on mount without route params', () => {
    renderComponent();

    expect(dispatchMock).toHaveBeenCalledTimes(0);
  });

  it('dispatches loadRepository on mount', () => {
    renderComponentWithRoute(`/${MOCK_REPOSITORY_1.owner}/${MOCK_REPOSITORY_1.name}`);

    expect(dispatchMock).toHaveBeenCalledWith(
      loadRepository({
        owner: MOCK_REPOSITORY_1.owner,
        name: MOCK_REPOSITORY_1.name,
        pagination: defaultExplorerPagination,
      }),
    );
  });

  it('shows loading spinner when loading and no repository', () => {
    renderComponent();

    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  it('renders repository info and issues when repository is present', () => {
    const mockRepo = {
      ...MOCK_REPOSITORY_1,
      issues: [MOCK_REPOSITORY_ISSUE_1, MOCK_REPOSITORY_ISSUE_2],
    };
    mockUseAppSelector.mockImplementation((selector) => {
      if (selector === selectRepository) {
        return mockRepo;
      }

      return undefined;
    });

    renderComponent();

    expect(screen.getByText(MOCK_REPOSITORY_ISSUE_1.title)).toBeInTheDocument();
    expect(screen.getByText(MOCK_REPOSITORY_ISSUE_2.title)).toBeInTheDocument();
    expect(screen.getByText(/Repositories/)).toBeInTheDocument();
  });

  it('renders empty state when not loading and no repository', () => {
    // Simulate explorerEvent error to set loading to false
    mockUseAppSelector.mockImplementation((selector) => {
      if (selector === selectExplorerEvent) {
        return MOCK_EVENT_LOAD_REPOSITORY_ERROR;
      }

      return undefined;
    });

    renderComponent();

    expect(screen.getByText(/No repository available/i)).toBeInTheDocument();
    expect(screen.getByText(/Please refresh page or try again later/i)).toBeInTheDocument();
    expect(screen.getByText(/Repositories/)).toBeInTheDocument();
  });

  it('dispatches loadRepository on pagination change', () => {
    const mockRepo = {
      ...MOCK_REPOSITORY_1,
      issues: [MOCK_REPOSITORY_ISSUE_1, MOCK_REPOSITORY_ISSUE_2],
    };
    mockUseAppSelector.mockImplementation((selector) => {
      if (selector === selectRepository) {
        return mockRepo;
      } else if (selector === selectPageInfo) {
        return MOCK_EXPLORER_PAGE_INFO_1;
      }

      return undefined;
    });

    renderComponentWithRoute(`/${MOCK_REPOSITORY_1.owner}/${MOCK_REPOSITORY_1.name}`);

    const dropdown = screen.getByTestId('pagination-dropdown');
    fireEvent.click(dropdown);
    const option = screen.getByText('10');
    fireEvent.click(option);

    expect(dispatchMock).toHaveBeenCalled();
  });
});
