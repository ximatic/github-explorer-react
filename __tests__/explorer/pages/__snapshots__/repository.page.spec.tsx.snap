// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="repositories h-full"
  >
    <div
      class="grid h-full w-full place-items-center"
    >
      <div
        aria-busy="true"
        class="p-progress-spinner"
        data-pc-name="progressspinner"
        data-pc-section="root"
        role="progressbar"
      >
        <svg
          class="p-progress-spinner-svg"
          data-pc-section="spinner"
          style="animation-duration: 2s;"
          viewBox="25 25 50 50"
        >
          <circle
            class="p-progress-spinner-circle"
            cx="50"
            cy="50"
            data-pc-section="circle"
            fill="none"
            r="20"
            stroke-miterlimit="10"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
